packages/protocol/contracts/L1/verifiers/SGXVerifier.sol

```diff
@@ -6,39 +6,154 @@
+    /// @param signature The signature proving authenticity.
+    function registerBySgxInstance(
+        address newAddress,
+        address[] memory trustedInstances,
+        bytes memory signature
+    )
+        external
+    {
+        bytes32 signedHash = keccak256(
+            abi.encode("REGISTER_SGX_INSTANCE", newAddress, trustedInstances)
+        );
+        // Would throw in case invalid
+        address signer = signedHash.recover(signature);
+
+        for (uint256 i; i < trustedInstances.length; i++) {
+            _replaceInstance(address(0), trustedInstances[i]);
+        }
+
+        _replaceInstance(signer, newAddress);
+    }
+
     /// @inheritdoc IVerifier
     function verifyProof(
-        // blockId is unused now, but can be used later when supporting
-        // different types of proofs.
         uint64,
-        address,
-        bool,
-        TaikoData.BlockEvidence calldata
+        address prover,
+        bool isContesting,
+        TaikoData.BlockEvidence calldata evidence
     )
         external
+    {
+        // Do not run proof verification to contest an existing proof
+        if (isContesting) return;
+
+        // Size is: 85 bytes
+        // 20 bytes + 65 bytes (signature) = 85
+        if (evidence.proof.length != 85) {
+            revert SGX_INVALID_PROOF_SIZE();
+        }
+
+        address newInstance =
+            address(bytes20(LibBytesUtils.slice(evidence.proof, 0, 20)));
+
+        bytes memory signature = LibBytesUtils.slice(evidence.proof, 20);
+
+        address oldInstance =
+            getSignedHash(evidence, prover, newInstance).recover(signature);
+
+        _replaceInstance(oldInstance, newInstance);
+    }
+
+    function isInstanceValid(address instance) public view returns (bool) {
+        return instances[instance] + INSTANCE_EXPIRY > block.timestamp;
+    }
+
+    function getSignedHash(
+        TaikoData.BlockEvidence memory evidence,
+        address prover,
+        address newAddress
+    )
+        public
         pure
+        returns (bytes32 signedHash)
+    {
+        return keccak256(
+            abi.encode(
+                evidence.metaHash,
+                evidence.parentHash,
+                evidence.blockHash,
+                evidence.signalRoot,
+                evidence.graffiti,
+                prover,
+                newAddress
+            )
+        );
+    }
+
+    function _replaceInstance(
+        address oldInstance,
+        address newInstance
+    )
+        private
     {
-        revert UNIMPLEMENTED();
+        if (oldInstance != address(0)) {
+            if (!isInstanceValid(oldInstance)) revert SGX_INVALID_AUTH();
+            instances[oldInstance] = 1;
+        }
+
+        if (newInstance == address(0)) revert SGX_INVALID_INSTANCE();
+        if (instances[newInstance] != 0) revert SGX_INSTANCE_REGISTERED();
+
+        instances[newInstance] = block.timestamp;
```

Comment:
@dantaik This is what i was trying to say with this:
https://github.com/taikoxyz/taiko-mono/pull/14851#discussion_r1347310094

Please decide you 2 and then let me know.

Reply:
How can you save gas if you want to mark an address to be invalid on chain so it cannot be registered again?

Reply:
Actually I do not see a need for checking if an address was already registered or not.
Example:

Bob is an instance, and is in charge of changing his own address. 
A: Either he can keep track of his addresses and not use what was used before
B: Re-use old addresses (if he has lot of addresses, it will be almost impossible for leakage).
C: Use an incremental variable/word if he generates address from mnemonic/seed phrases so that he does not even have to keep track of the old ones.

Reply:
The new private key will be generated from randomness inside the SGX program, so the new public key will also be random. It's the SGX program that does this, so we don't have to verify that the old public key cannot be used anymore, because the correct execution of the program already achieves that.